pyquil/api/_qpu.py

```diff
@@ -154,3 +182,63 @@ def _build_patch_values(self) -> dict:
             patch_table[k.name][k.index] = v
 
         return patch_table
+
+    def _build_arithmetic_patch_values(self) -> str:
+        """
+        Update the variables shim with the final values to be patched into the gate parameters,
+        according to the arithmetic expressions in the original program.
+
+        For example:
+
+            DECLARE theta REAL
+            RZ(3 * theta) 0
+
+        gets translated to:
+
+            DECLARE theta REAL
+            DECLARE __P REAL[0]
+            RZ(__P[0]) 0
+
+        and the recalculation table gets the entry: __P[0] = '3 * theta'. In our example, we
+        look up the value of 'theta' in the variables shim, multiply that by 3, and then add
+        an entry for __P[0] to the variables shim.
+
+        :return: Name of the region used to patch the parameters.
+        """
+        try:
+            recalculation_table = self._executable.recalculation_table
+        except AttributeError:
+            # No recalculation table, no work to be done here.
+            return ""
+
+        # We only declare one region to hold these values, and we need to add that name to
+        memory_reference_name = None
+        for memory_reference, expression in recalculation_table.items():
+            # Replace the memory references with any values the user has written
+            value = self._resolve_memory_references(expression)
+            self._variables_shim[memory_reference] = value
+            memory_reference_name = memory_reference.name
+        return memory_reference_name
```

Comment:
wait, why are we returning the last memory reference name

Reply:
also sometimes you return "" and sometimes you return `None`

Reply:
This is a good question. I'm not sure what the best solution is here, I will message you on slack

Reply:
(The "" vs None I've fixed locally, thanks for noticing)

Reply:
```python
memory_reference_name = list(set(mr.name for mr in recalculation_table.keys()))
assert len(memory_reference_name) == 1, 'meaningful error message'
memory_reference_name = memory_refrence_name[0]

for mr, exp in rec_table.items():
  var_shim[mr] = resolve_mem_ref(exp)

return memory_reference_name
```